test/jdk/com/sun/jdi/JdwpOnThrowTest.java

```diff
@@ -0,0 +1,161 @@
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import com.sun.jdi.Bootstrap;
+import com.sun.jdi.VirtualMachine;
+import com.sun.jdi.connect.AttachingConnector;
+import com.sun.jdi.connect.Connector;
+import com.sun.jdi.connect.IllegalConnectorArgumentsException;
+import com.sun.jdi.connect.ListeningConnector;
+import com.sun.jdi.event.EventIterator;
+import com.sun.jdi.event.EventQueue;
+import com.sun.jdi.event.EventSet;
+import com.sun.jdi.event.Event;
+import com.sun.jdi.event.ExceptionEvent;
+import com.sun.jdi.request.EventRequestManager;
+import jdk.test.lib.Utils;
+import lib.jdb.Debuggee;
+
+import java.io.IOException;
+import java.net.Inet4Address;
+import java.net.Inet6Address;
+import java.net.InetAddress;
+import java.net.ServerSocket;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+/*
+ * @test
+ * @bug 8317920
+ * @summary Tests for JDWP agent to send valid exception event with onthrow option
+ * @library /test/lib
+ *
+ * @build ThrowCaughtException JdwpOnThrowTest
+ * @run main/othervm JdwpOnThrowTest
+ */
+public class JdwpOnThrowTest {
+
+    private static long TIMEOUT = 10000;
+
+    public static void main(String[] args) throws Exception {
+        int port = findFreePort();
+        try (Debuggee debuggee = Debuggee.launcher("ThrowCaughtException").setAddress("localhost:" + port)
+                                         .enableOnThrow("Ex", "Start").setSuspended(true).launch()) {
+            VirtualMachine vm = null;
+            try {
+                vm = attach("localhost", "" + port);
+                EventQueue queue = vm.eventQueue();
+                log("Waiting for exception event");
+                long start = System.currentTimeMillis();
+                while (start + TIMEOUT > System.currentTimeMillis()) {
+                    EventSet eventSet = queue.remove(TIMEOUT);
+                    EventIterator eventIterator = eventSet.eventIterator();
+                    while(eventIterator.hasNext() && start + TIMEOUT > System.currentTimeMillis()) {
+                        Event event = eventIterator.next();
+                        if (event instanceof ExceptionEvent ex) {
```

Comment:
It's a bit unclear to me why this test is even seeing the ExceptionEvent since I don't see any place where an ExceptionEventRequest has been setup.

Reply:
It seems that ex.request()) is null for this event. The spec for Event.request() says null might happen for VMDeath events, but doesn't mention any other. In any case, I guess JDI does not filter out any events for which there is not a request, which is why the test is seeing it. This also means the event could be considered unexpected if the debugger did not explicitly add an ExceptionEventRequest, and it could in fact cause an NPE if the debugger decides to look at ex.request().

On the other hand, I think if the debug agent requires that a handler be in place (as a result of the debugger creating and enabling the EventRequest) in order for onthrow to send the event, then the handler creation might come too late, and the event will get filtered out.

Reply:
Yes, so the current way is probably ok?

IDEs like IntelliJ properly handle the exception event that is sent and don't report an error, so I don't see a problem.

Reply:
> On the other hand, I think if the debug agent requires that a handler be in place (as a result of the debugger creating and enabling the EventRequest) in order for onthrow to send the event, then the handler creation might come too late, and the event will get filtered out.

I was thinking that maybe suspend=y is the way around this for the debugger. This will cause the debug agent to do a SUSPEND_ALL before sending the ExceptionEvent. However, the event could still be received by the debugger before the debugger has setup the ExceptionRequest. The only thing suspend=y is really doing here is ensuring that all threads get suspended when the event is generated. In fact that makes it even more important that the debugger process the event. Otherwise it won't be doing eventSet.resume(), and execution will come to a stand still.

I think we might just be stuck with what we have now. My main concern then is that it is neither spec'd nor documented well. I'll file a separate CR to cover that.

Reply:
I'm happy to help with any documentation efforts, as the lack of proper documentation is probably one of the reasons why the adoption was so low over the last twenty years.

The documentation should tell the potential users about the possible issues with this option.